lib/remote/apilistener-configsync.cpp

```diff
@@ -325,27 +321,26 @@ void ApiListener::UpdateConfigObject(const ConfigObject::Ptr& object, const Mess
 	params->Set("name", object->GetName());
 	params->Set("type", object->GetReflectionType()->GetName());
 	params->Set("version", object->GetVersion());
+	params->Set("package", object->GetPackage());
 
 	String zoneName = object->GetZoneName();
 
 	if (!zoneName.IsEmpty())
 		params->Set("zone", zoneName);
 
-	if (object->GetPackage() == "_api") {
-		String file;
-
-		try {
-			file = ConfigObjectUtility::GetObjectConfigPath(object->GetReflectionType(), object->GetName());
-		} catch (const std::exception& ex) {
-			Log(LogNotice, "ApiListener")
-				<< "Cannot sync object '" << object->GetName() << "': " << ex.what();
-			return;
-		}
+	String file;
 
-		std::ifstream fp(file.CStr(), std::ifstream::binary);
-		if (!fp)
-			return;
+	try {
+		file = ConfigObjectUtility::GetObjectConfigPath(object->GetReflectionType(), object->GetName(),
+			object->GetPackage());
+	} catch (const std::exception& ex) {
+		Log(LogNotice, "ApiListener")
+			<< "Cannot sync object '" << object->GetName() << "': " << ex.what();
+		return;
+	}
 
+	std::ifstream fp(file.CStr(), std::ifstream::binary);
+	if (fp) {
 		String content((std::istreambuf_iterator<char>(fp)), std::istreambuf_iterator<char>());
 		params->Set("config", content);
 	}
```

Comment:
I don't understand this change. If the object is not the `_api` package, there's no guarantee that this file exists anyways. And if it's in the `_api` package, this changes the behavior when there's an error opening the file.

Reply:
With this change we support objects created via the API and objects created via the Director.

The config is used [here](https://github.com/WuerthPhoenix/icinga2/blob/5aacae705fc17d8429b2e1e6d146ae9e218f3cee/lib/remote/apilistener-configsync.cpp#L117) in `ConfigUpdateObjectAPIHandler` in conjunction with the object to figure out if it is a new object.

If it is a new object created via API, the file will be present. If the object was created with the Director, when it is updated via API the file will not be found (the director uses different filenames) but the object will not be empty so the changes will be applied.

For sure if the file exists and an error occurs while reading, it will no longer fail [here](https://github.com/WuerthPhoenix/icinga2/blob/5aacae705fc17d8429b2e1e6d146ae9e218f3cee/lib/remote/apilistener-configsync.cpp#L343) but fail [here](https://github.com/WuerthPhoenix/icinga2/blob/5aacae705fc17d8429b2e1e6d146ae9e218f3cee/lib/remote/apilistener-configsync.cpp#L149).

Do you suggest adding something like this in order to keep the same behavior for the _api package?

```
std::ifstream fp(file.CStr(), std::ifstream::binary);
if (fp) {
	String content((std::istreambuf_iterator<char>(fp)), std::istreambuf_iterator<char>());
	params->Set("config", content);
} else if (!fp && object->GetPackage() == "_api") {
	return;
}
```